infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -131,60 +142,126 @@ public LogUnitServer(ServerContext serverContext) {
      */
     @ServerHandler(type = CorfuMsgType.TAIL_REQUEST)
     public void handleTailRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        TailsResponse tails = batchWriter.queryTails(msg.getEpoch());
-        r.sendResponse(ctx, msg, CorfuMsgType.TAIL_RESPONSE.payloadMsg(tails));
+        log.debug("handleTailRequest: received a tail request {}", msg);
+        CorfuPayloadMsg<Void> tailsReqMsg = new CorfuPayloadMsg<>();
+        tailsReqMsg.copyBaseFields(msg);
+        batchWriter.<TailsResponse>addTask(TAILS_QUERY, tailsReqMsg)
```

Comment:
Question: why tails query needs to go to batch writer?

Reply:
The `StreamLog::getTails` is not thread safe, so it is "made" thread safe by executing it sequentially with respect to other writes.